### ISSUE TITLE
Thread Sanitizer complains about a data race

### ISSUE THREAD

**Issue by scosu** [benjamin-bader/cppmetrics#9]
I just noticed, that the thread sanitizer complains about a data race between modify() and count() due to the dynamic allocation of the cells:

```
WARNING: ThreadSanitizer: data race (pid=1078583)
  Read of size 8 at 0x7b500015d080 by thread T43:
    #0 cppmetrics::LongAdder::Count() const cppmetrics/LongAdder.cc:152:9

  Previous write of size 8 at 0x7b500015d080 by thread T8:
    #0 cppmetrics::LongAdder::modify(long) cppmetrics/LongAdder.cc:186:28
    #1 cppmetrics::LongAdder::Increment(long) cppmetrics/LongAdder.cc:138:3

  Location is heap block of size 512 at 0x7b500015d000 allocated by main thread:
    #0 operator new(unsigned long) /tmp/builder/llvm/projects/compiler-rt/lib/tsan/rtl/tsan_new_delete.cpp:64:3
    #1 __gnu_cxx::new_allocator<cppmetrics::LongAdder::Cell*>::allocate(unsigned long, void const*) /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/ext/new_allocator.h:104:27
    #2 std::_Vector_base<cppmetrics::LongAdder::Cell*, std::allocator<cppmetrics::LongAdder::Cell*> >::_M_allocate(unsigned long) /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_vector.h:168:35
    #3 std::vector<cppmetrics::LongAdder::Cell*, std::allocator<cppmetrics::LongAdder::Cell*> >::_M_fill_insert(__gnu_cxx::__normal_iterator<cppmetrics::LongAdder::Cell**, std::vector<cppmetrics::LongAdder::Cell*, std::allocator<cppmetrics::LongAdder::Cell*> > >, unsigned long, cppmetrics::LongAdder::Cell* const&) /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/vector.tcc:483:34
    #4 std::vector<cppmetrics::LongAdder::Cell*, std::allocator<cppmetrics::LongAdder::Cell*> >::insert(__gnu_cxx::__normal_iterator<cppmetrics::LongAdder::Cell**, std::vector<cppmetrics::LongAdder::Cell*, std::allocator<cppmetrics::LongAdder::Cell*> > >, unsigned long, cppmetrics::LongAdder::Cell* const&) /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_vector.h:1024:9
    #5 std::vector<cppmetrics::LongAdder::Cell*, std::allocator<cppmetrics::LongAdder::Cell*> >::resize(unsigned long, cppmetrics::LongAdder::Cell* const&) /usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_vector.h:687:4
    #6 cppmetrics::LongAdder::LongAdder() cppmetrics/LongAdder.cc:122:11
```

I shortened the stacks to the LongAdder code.

Maybe the cell pointers have to be atomic? Or is it correct to assume that the pointer is always atomically written/read?

**Comment by benjamin-bader** [comment]
I haven't seen this yet (and honestly haven't used this code in a few years), so please take this all with a grain of salt.

`count()` is not guaranteed in the first place to see all concurrent modifications: https://github.com/benjamin-bader/cppmetrics/blob/master/metrics/include/metrics/LongAdder.h#L36.  That being said -

I _believe_ pointer reads/writes are atomic given that they are word-sized and, I think, are all aligned.  That maybe a mistaken belief, particularly on non-x86_64 architectures!  Perhaps there's the _other_ sense of atomicity that's wrong here, where threads on other cores won't see the same allocations.  I'm somewhat confident that the existing locking scheme prevents this via memory barriers, but don't ask me to write a proof 😬 

TL;DR: A few years back I was convinced that this code was thread-safe.  If you or your tools find otherwise, I'll be grateful to know as much as you can tell me about it!

**Comment by scosu** [comment]
It is just Thread Sanitizer that complains about it. I *think* the code is thread safe as well.

On the other hand there is a C++ memory ordering model that exactly ensures only atomicitiy of an operation and nothing else, see https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering memory_order_relaxed . So maybe it would be good to have a vector of atomic pointers that are stored and loaded with memory_order_relaxed to ensure they are atomic operations?
